### PR TITLE
Add an error type to indicate a message should be deferred

### PR DESCRIPTION
**Pull request by zhedahht**
Currently a message is completed when the listener message handler finishes without error, and it's abandoned when the handler meets an error.
 
We have scenarios that messages have to be deferred before they are not ready to be handled. Let's define a specific error type to indicate that a message is going to be deferred.

### CODE REVIEW

File: listener.go [hunk 1]
```diff
@@ -14,6 +14,14 @@ const (
 	defaultSubscriptionName = "default"
 )
 
+// DeferMessageError is an error type to indicate a message is to be deferred
+type DeferMessageError struct{}
```
Comment (paulgmiller) [inline listener.go:18]: This is similiar to my old strawman request here
https://github.com/keikumata/azure-pub-sub/pull/13/files

Kind of liked the interface so we could say how long to defer but okay with this too.

File: listener.go [hunk 2]
```diff
@@ -175,15 +183,18 @@ func (l *Listener) Listen(ctx context.Context, handle Handle, topicName string,
 			if val, ok := message.UserProperties["type"]; ok {
 				err := handle(ctx, string(message.Data), val.(string))
 				if err != nil {
-					err = message.Abandon(ctx)
-					return err
+					_, ok := err.(*DeferMessageError)
+					if ok {
+						return message.Defer(ctx)
```
Comment (paulgmiller) [inline listener.go:188]: what's the default length of defer?

File: listener.go [hunk 3]
```diff
@@ -14,6 +14,14 @@ const (
 	defaultSubscriptionName = "default"
 )
 
+// DeferMessageError is an error type to indicate a message is to be deferred
+type DeferMessageError struct{}
+
+// Error converts the error into a string
+func (dfe *DeferMessageError) Error() string {
+	return ""
```
Comment (paulgmiller) [inline listener.go:22]: "Can't handle messsage currrently... deferring...."?

**Review by paulgmiller — APPROVED** [review]

**Comment by paulgmiller** [comment]
Approved but obviously want Kei or Stephane feedback (Stephane is on france time now so probably not till tommorow :)

**Comment by serbrech** [comment]
I don't think this does what you think it does.
documentation is here : https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-deferral#message-deferral-apis

Defer kicks the message onto a defer queue, and it's not automatically pushed back to you.

The easiest way to handle this is to basically defer it in-memory
This is how the .net SDK implements it : [Task.Delay](https://github.com/Azure/azure-sdk-for-net/blob/1e24500806dc01be9e81cb9106c397d17df4512c/sdk/servicebus/Microsoft.Azure.ServiceBus/src/RetryPolicy.cs#L103)

I agree that we need to support this as part of the tooling and not ask each service to implement it. especially when it comes to exponential backoff. It should arguably be part of the sdk, but they are very understaffed. maybe we can port it back there once we have it here.
Since the message won't disappear from the queue until you explicitely acknowledge it, we can just hold it in memory and push it back for processing after a given delay. if the app crashes, we're fine.
I think that's the easiest solution. it also prevents the message going to the `dead letter queue` because you reach max retry.
reching max retry. your case is not an error that you want to eventually discard. it's just saying wait.

The other solution is what Paul did, but that also isn't great IMO.
You are a receiver, and you are not supposed to *write* to the topic.
if you push the message back onto the topic as a scheduled message, you become an emitter. the message you emit is a NEW message, not the same, and if dedup is enabled, and you don't change the messageID, Servicebus will just swallow your scheduled message as a duplicate.

**Comment by serbrech** [comment]
closing in favor of #23

**Pull request closed without merging**